sys/chunked_bucket.js

```diff
@@ -0,0 +1,433 @@
+"use strict";
+
+var P = require('bluebird');
+var uuid = require('cassandra-uuid').TimeUuid;
+var mwUtil = require('../lib/mwUtil');
+var HTTPError = require('../lib/exports').HTTPError;
+var URI = require('swagger-router').URI;
+var preq = require('preq');
+
+var yaml = require('js-yaml');
+var fs = require('fs');
+var spec = yaml.safeLoad(fs.readFileSync(__dirname + '/chunked_bucket.yaml'));
+
+/**
+ * The chunk size to use for data slicing
+ * @type {number}
+ * @const
+ */
+var CHUNK_SIZE = 31000;
+
+/**
+ * The grace_ttl parameter for a revision policy
+ * @type {number}
+ * @const
+ */
+var GRACE_TTL = 86400;
+
+function range(N) {
+    return Array.apply(null, { length: N }).map(Number.call, Number);
+}
+
+function ChunkedBucket(options) {
+    this.log = options.log || function() {};
+}
+
+ChunkedBucket.prototype._sliceData = function(data) {
+    var slicer;
+    if (Buffer.isBuffer(data)) {
+        slicer = Buffer.prototype.slice;
+    } else {
+        data = data.toString();
+        slicer = String.prototype.substring;
+    }
+    var result = [];
+    for (var index = 0; index < data.length; index += CHUNK_SIZE) {
+        result.push(slicer.call(data, index, index + CHUNK_SIZE));
+    }
+    return result;
+};
+
+ChunkedBucket.prototype._joinData = function(dataParts) {
+    if (!dataParts || !dataParts.length) {
+        return '';
+    } else if (Buffer.isBuffer(dataParts[0])) {
+        return Buffer.concat(dataParts);
+    } else {
+        return String.prototype.concat.apply(dataParts[0], dataParts.slice(1));
```

Comment:
This is equivalent to '`'.concat(dataParts)`, with the latter avoiding the slice.